PIG 6 - CTA observation handling

dcfidalgo opened the pull request:
:pig_nose: :pig_nose: :pig_nose: 
In this :pig: we outline a new scheme for the data handling in Gammapy.
Discussions about it started during the Coding Sprint in Madrid last week.

Here i uploaded a first advanced draft of the PIG, which is mainly based on discussions/suggestions by @cdeil and @registerrier. Maybe we can go over it tomorrow in the call and discuss the next steps.

Once the PIG has reached its final version, i will write it down in the *rst* format so it can go to the docs.
For now i find it more convenient to write it in a notebook as markdown.
David

cdeil commented on docs/development/pigs/pig-006.md:
```diff
@@ -0,0 +1,114 @@
+
+# PIG 6 - Data/GTI handling in Gammapy
```
@dcfidalgo - Thanks! I'll have a look soon.

Can we have the PIG and PR title be identical?
Maybe "CTA DL3 data handling in Gammapy" or "CTA observation handling in Gammapy"?

I mean, it would be desirable to also have a PIG / plan for slewing telescope observation handling (Fermi-LAT and HAWC, maybe CTA), and to describe all "data handling" along the analysis chain, i.e. how the DL3 EVENTS, IRFs and AUX data goes on as binned maps and spectra and reduced IRFs.

But then this PIG would never finish. I think the scope and content you chose here is good. I'm just suggesting you clarify this a little more in the title and abstract.

dcfidalgo replied:
Ok, i updated the title and made it the same for the PIG and PR.
The abstract now also states more clearly that the focus lies merely on resolving current limitations when analysing IACT data.

cdeil commented on docs/development/pigs/pig-006.md:
```diff
@@ -0,0 +1,114 @@
+  * has a unique identifier, the *obs_id*
+  * contains 1 event list, 1 GTI table and 1 corresponding set of IRFs (aeff, edisp, psf)
+  * has 1 dead-time fraction (`DEADC`)
+  * the live time of the observation is obtained as `ONTIME * (1 - DEADC)`
```
Can you try to flesh out the GTI part of the PIG a bit?
You know, be more clear what we assume about GTI and how we use it, to have a more focused discussion.

I think we can make GTI required and fully rely on it being there.
Maybe, if easy, we could allow loading data without a GTI if TSTART and TSTOP is given in the EVENTS header, but then IMO the first step should be to create a GTI table, not try to drag that alternative format to have run times along in Gammapy.

Why have GTI at all and not just START and STOP for an observation?
No strong reason, it's just a little more flexible, e.g. if there's 5 seconds of hardware problems or a meteor, it can be cut out without having to start a new EVENTS and observation, if some IACT wants that.
It's also closer then to the data model that Fermi-LAT uses already, they absolutely require GTI.

My thinking for DEADC is that it should go away:
https://github.com/open-gamma-ray-astro/gamma-astro-data-formats/issues/62#issuecomment-428221596
For now, we should keep it, because some people might use it.
But I'd suggest linking to that issue from the PIG and trying to design it as much as possible that DEADC is an implementation detail that can be removed later without having to touch high-level code.

For PHASE, you could also try to be a bit more specific. For me, the key point in this PIG is to solve the use case of GTI and PHASE, and to defer other things like EVENT_TYPE to a future PIG. Suggest you state that explicitly, that this will not be solved here, but next year in a new PIG. But for PHASE, I think it's easy to solve in a generic solution like for GTI: add a table with PHASE START / STOP intervals, to allow a user to run an analysis with any PHASE selection they like. Do you think it's a good idea, or do you want to not develop the full solution how to do PHASE in this PIG?

Final comment: I'm not sure having the PIG in markdown is the best solution. RST is almost the same, and there shouldn't be duplication of content in one document and another notebook to avoid confusion / extra work during the review and discussion in the coming days (weeks?). Also version history will be lost when you rename MD to RST. So suggest to change to RST now. But it's not a big point, if you really want to do the review / discussion with MD and change later, OK.

kosack commented on docs/development/pigs/pig-006.md:
```diff
@@ -0,0 +1,116 @@
+
+* **Observation**:
+  * has a unique identifier, the *obs_id*
+  * contains 1 event list, 1 GTI table and 1 corresponding set of IRFs (aeff, edisp, psf)
```
note that the IRFs are associated to a GTI, not to the observation in general. E.g. you might have a single GTI, and then it's as we have now, or you might have 2 or more GTIs in the observation, where the calibration is quite different (e.g. a GTI with perfect conditions, and several others others with cloudy but still usable conditions).

cdeil replied:
@kosack - Both Gammapy and ctools currently assume one IRF per observation.

To quote from http://cta.irap.omp.eu/gammalib/users/user_manual/modules/obs.html

> A gamma-ray observation is defined for a single specific instrument, and describes a time period during which the instrument is in a given stable configuration that can be characterized by a single specific response function. Each gamma-ray observation is composed of events and a response function.

If you want to distribute one IRF per GTI, we can implement that in Gammapy in the coming months, but it's a big decision, not what was assumed from the science tools so far.

That said, we'll probably anyways have to introduce a major change to support EVENT_TYPE, where you have one IRF per OBS_ID and EVENT_TYPE.

One option on the Gammapy side could be to stick with one IRF per OBS_ID for now and just to introduce the EVENT and GTI filter functionality, work with you and CTA to define the new DL3 model, and once there is a concrete proposal, prototype that early next year.

dcfidalgo commented on docs/development/pigs/pig-006.rst:
```diff
@@ -0,0 +1,137 @@
+   - has a unique identifier, the *obs_id*
+   - contains 1 event list, 1 GTI table and 1 corresponding set of IRFs (aeff, edisp, psf)
+   - has 1 dead-time fraction (`DEADC`)
+   - the live time of the observation is obtained as `ONTIME * (1 - DEADC)`
```
By @cdeil:
> Can you try to flesh out the GTI part of the PIG a bit?
> You know, be more clear what we assume about GTI and how we use it, to have a more focused discussion.
> 
> I think we can make GTI required and fully rely on it being there.
> Maybe, if easy, we could allow loading data without a GTI if TSTART and TSTOP is given in the EVENTS header, but then IMO the first step should be to create a GTI table, not try to drag that alternative format to have run times along in Gammapy.
> 
> Why have GTI at all and not just START and STOP for an observation?
> No strong reason, it's just a little more flexible, e.g. if there's 5 seconds of hardware problems or a meteor, it can be cut out without having to start a new EVENTS and observation, if some IACT wants that.
> It's also closer then to the data model that Fermi-LAT uses already, they absolutely require GTI.
> 
> My thinking for DEADC is that it should go away:
> https://github.com/open-gamma-ray-astro/gamma-astro-data-formats/issues/62#issuecomment-428221596
> For now, we should keep it, because some people might use it.
> But I'd suggest linking to that issue from the PIG and trying to design it as much as possible that DEADC is an implementation detail that can be removed later without having to touch high-level code.
> 
> For PHASE, you could also try to be a bit more specific. For me, the key point in this PIG is to solve the use case of GTI and PHASE, and to defer other things like EVENT_TYPE to a future PIG. Suggest you state that explicitly, that this will not be solved here, but next year in a new PIG. But for PHASE, I think it's easy to solve in a generic solution like for GTI: add a table with PHASE START / STOP intervals, to allow a user to run an analysis with any PHASE selection they like. Do you think it's a good idea, or do you want to not develop the full solution how to do PHASE in this PIG?
> 
> Final comment: I'm not sure having the PIG in markdown is the best solution. RST is almost the same, and there shouldn't be duplication of content in one document and another notebook to avoid confusion / extra work during the review and discussion in the coming days (weeks?). Also version history will be lost when you rename MD to RST. So suggest to change to RST now. But it's not a big point, if you really want to do the review / discussion with MD and change later, OK.
>

dcfidalgo replied:
Stated that we require GTI tables, that they are more flexible than simple STAR STOP tuples, linked to the `DEADC` discussion

dcfidalgo commented:
I just moved to the rst format, and i just copied the inline comments made so far ... i did not know how to preserve them otherwise.

Thanks @cdeil for your extensive feedback!

> For the PIG document, I feel like the separation into this document and https://github.com/gammapy/gammapy-extra/blob/master/experiments/data.ipynb is problematic. I read both just now, and already I'm not sure which one to comment on exactly. Would strongly prefer if you put all content to be discussed / reviewed in the one document here in the coming days, and already use RST like we do for all PIGs.

I deleted (commented out for now) the reference to the `data.ipynb`, so this is not a part of the PIG anymore. Maybe i will copy some relevant parts from there to the rst document.

> common use cases, either via text or code snippets, that will evolve into automated tests. These should be 3-10 lines long, i.e. less code than in the notebook currently. See https://behave.readthedocs.io/en/latest/tutorial.html as examples for one way you can structure them, using "use cases", "scenarios" and "given - when - then".

Nice, i will add some scenarios in this format!

> rough steps how to implement them. Like I said, I would do the change to Observations first, and then probably I would already introduce the ObservationFilter class on the side, and then step by step hook it up to the Observation and Observations API to allow the use cases defined. A first one could be to load some data like we do now with the datastore, do a GTI and / or EVENTS selection, and write out the transformed data to a new folder. A second one could be to load up CTA IRFs, set some obs parameters like pointing and time via the API, simulate some events (can just be a zero-filled table for now, event simulator will come later), and write that to disk or pass it to an analysis.

+1 Sounds good, will add this asap.

> If your diagram, you are missing "edisp" as a field in Observation.

Ok, i will produce a more complete diagram. For now it was merely indicative and i actually missed a lot of `...` in the class attributes and methods.

> The key question of where observation metadata like pointing position, and times comes from isn't specified yet. In the old scheme they come from the index tables. Where do they come from / when are they read in the new scheme?

For now i would stick to the old scheme (except computing the observation time from the GTI table). Maybe it makes more sense to think about a new scheme once we have more details how the *observation library* in CTA will look like. To support ctool's XML format for now, we could maybe create the index table on the fly like you suggested.

> But it should be stated explicitly that the design and API will change again in 2019. And we probably shouldn't spend too much effort now in this first phase on trying to get the API and design right, because we won't without prototyping EVENT_TYPE. So let's just evolve Gammapy a bit, to support the GTI and filter feature?
 
I totally agree! I think for coming up with a clean and lasting scheme of the data handling we need more specifics and final requirements of the definite data format of CTA. I will state this more clearly in the PIG.

> Basically I think there need to be two levels of information: always load "observation metadata" and have things like pointing position and start / stop loaded and stored on the observation object (and probably duplicated on the observations and/or datastore objects as tables for obs selection.

In our current scheme, the `DataStoreObservation` does not load or hold any data on creation, except for the obs_id. Instead it *knows* where to find the information via the `DataStore`object and its obs_id, and the "observation metadata" is only stored once it is explicitly accessed. Like i said, i do not plan to change much on this scheme just because it will most likely change in 2019.

> And then for events (and probably IRF), support both in-memory and proxy to on-disk. In your diagram you have this distinction on the DataStore classes. I think my main design question is: shouldn't this be on the Observation class instead? I mean: it's not either / or for the datastore: observation metadata and some things lilke filtered EVENTS need to be in-memory, while still EVENTS and IRFs from potentially 1000s of other observations remain on-disk.

The idea of the filters is to apply them on-the-fly, so in principle we do not need to store filtered event lists in memory.

So i think to achieve following three things:
* select observations by time
* select observations by phase
* allow observations to hold simulated event lists

the shortest and easiest way is to:

* evolve the `ObservationList` class into a more user friendly `Observations` class
* evolve the `Observation` class from the `DataStoreObservation` class, keeping it a pure proxy class, but with a filter
* add a new `DataStore` class that can hold event lists in memory

Maybe it is not the ultimate new super data handling scheme that we hoped for :), but i think we all agree on that it will change in 2019, maybe in a very profound way (like looping over time intervals instead of observations), so not too much time should be spent on this now.

cdeil commented:
@adonath @registerrier and I as Gammapy lead developers have decided to close out this old PR and discussion by withdrawing this PIG. See 4694f471a92868de793b78dacecd330c510f2117

The last comment here is from Oct 2018, over a year ago, and it's unlikely that someone will pick up and carry forward this PIG to a state where it can be circulated and approved or rejected. Instead work on DL3 data handling and CTA support is continuing and will continue in the coming years in individual pull requests, new future PIGs, and within CTA (requirements, specs, ...).

Like all PIGs, this one will appear at https://docs.gammapy.org/0.14/development/pigs/index.html and be forever available as a reference, closing out PIGs like this doesn't mean that the work is lost. Thank you to everyone that contributed here!
